[Feature#1270][Connectors] add source plugin: seatunnel-connector-spark-http

tmljob opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist

  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/incubator-seatunnel/issues).

  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.

  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.

-->

## Purpose of this pull request

<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/development/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

tmljob committed:
Merge pull request #1 from apache/dev

sync change from source project

tmljob committed:
[Feature#1270][Connectors] add plugin: seatunnel-connector-http

wuchunfu commented on seatunnel-connectors/seatunnel-connectors-spark/seatunnel-connector-spark-http/src/main/resources/META-INF/services/org.apache.seatunnel.spark.BaseSparkSource:
```diff
@@ -0,0 +1 @@
+org.apache.seatunnel.spark.source.Http
```
@tmljob  Here to add the license header, thanks

tmljob committed:
add license header

wuchunfu reviewed (commented):
@tmljob Please fix unit tests errors, thanks.

tmljob commented:
> @tmljob Please fix unit tests errors, thanks.

This ut has nothing to do with the code submitted by my pr. This is a problem that existed when fork from the dev branch. Do you need me to modify it here?

ruanwenjun commented:
> > @tmljob Please fix unit tests errors, thanks.
> 
> This ut has nothing to do with the code submitted by my pr. This is a problem that existed when fork from the dev branch. Do you need me to modify it here?

@tmljob The ut case should be fixed by upstream, could you please rebase the upstream/dev?

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-spark/seatunnel-connector-spark-http/src/main/java/org/apache/seatunnel/spark/source/Http.java:
```diff
@@ -0,0 +1,142 @@
+ * limitations under the License.
+ */
+
+package org.apache.seatunnel.spark.source;
```
It's better to change the package name to `org.apache.seatunnel.spark.http.source`.

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-spark/seatunnel-connector-spark-http/src/main/java/org/apache/seatunnel/spark/source/Http.java:
```diff
@@ -0,0 +1,142 @@
+    public Dataset<Row> getData(SparkEnvironment env) {
+        SparkSession spark = env.getSparkSession();
+        String url = config.getString(Settings.SOURCE_HTTP_URL);
+        String method = config.hasPath(Settings.SOURCE_HTTP_METHOD) ?
+            config.getString(Settings.SOURCE_HTTP_METHOD) : GET;
+        String header = config.hasPath(Settings.SOURCE_HTTP_HEADER) ?
+            config.getString(Settings.SOURCE_HTTP_HEADER) : "";
+        String requestParams = config.hasPath(Settings.SOURCE_HTTP_REQUEST_PARAMS) ?
+            config.getString(Settings.SOURCE_HTTP_REQUEST_PARAMS) : "";
+        String syncPath = config.hasPath(Settings.SOURCE_HTTP_SYNC_PATH) ?
+            config.getString(Settings.SOURCE_HTTP_SYNC_PATH) : "";
```
It's better to use `TypesafeConfigUtils#getConfig(final Config config, final String configKey, final T defaultValue)` , this is more readable.
e.g.
```java
TypesafeConfigUtils.getConfig(config, Settings.SOURCE_HTTP_METHOD, GET);
```

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-spark/seatunnel-connector-spark-http/src/main/java/org/apache/seatunnel/spark/source/Http.java:
```diff
@@ -0,0 +1,142 @@
+        try {
+            return new ObjectMapper().readValue(content, HashMap.class);
+        } catch (IOException e) {
+            LOG.error("{} json to map error!", content, e);
```
Why catch this exception, make this case to be normal? I think in most case, if you think this exception just need to be resolved in this method, you need to add a comment why this exception needn't to throw.

tmljob replied:
Catching this exception only records the log, does not handle it, and does not affect the main process.

ruanwenjun replied:
You need to add comment when you catch a exception and kill the exception in this catch.

tmljob replied:
Annotation has been added.

ruanwenjun reviewed (changes requested):
Good patch, I don't review all code, here are some suggestions.
And you need to add documentation about how to use this plugin.

tmljob commented:
> > > @tmljob Please fix unit tests errors, thanks.
> > 
> > 
> > This ut has nothing to do with the code submitted by my pr. This is a problem that existed when fork from the dev branch. Do you need me to modify it here?
> 
> @tmljob The ut case should be fixed by upstream, could you please rebase the upstream/dev?

I've synced the latest code from the dev branch, and this ut use case runs fine in my local environment alone.

tmljob committed:
fix review problem:
1、add document of http source plugin;
2、modify code to fix problems by reviewed;

tmljob committed:
format this pom.xml

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-spark/pom.xml:
```diff
@@ -18,40 +18,41 @@
-        <version>${revision}</version>
-    </parent>
-    <modelVersion>4.0.0</modelVersion>
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
It seems our code style doesn't cover the of pom, but you don't need to do this change, it's better to revert this change.

tmljob replied:
ok

ruanwenjun reviewed (changes requested):
Most look good to me, just a code style need to change.

tmljob committed:
Revert "format this pom.xml"

This reverts commit a7a094e76892abc249c338675b1c30cdc2155914.

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-spark/pom.xml:
```diff
@@ -52,6 +52,7 @@
         <module>seatunnel-connector-spark-neo4j</module>
         <module>seatunnel-connector-spark-iceberg</module>
         <module>seatunnel-connector-spark-feishu</module>
+		<module>seatunnel-connector-spark-http</module>
```
The format seems mismatching.

tmljob committed:
align new module

ruanwenjun reviewed (approved):
LGTM

wuchunfu reviewed (approved):
LGTM
